### ISSUE TITLE
Forecasting wish list

### ISSUE THREAD

**Issue by mloning** [alan-turing-institute/sktime#220]
# Forecasting wish list

### Atomic 
- [ ] DummyForecaster (strategies={"last", "seasonal-last", "mean"})
- [ ] TrendForecaster (linear, polynomial, etc)
- [ ] ThetaForecaster
- [ ] ExponentialSmoothingForecaster
- [ ] AutoARIMAForecaster

### Reduction
- [ ] ReducedRegressionForecaster
- [ ] ReducedTimeSeriesRegressionForecaster

## Transformers
- [ ] Detrender (series to series, meta-estimator that works with any forecaster)
- [ ] Imputer

## Composition
- [ ] TransformedTargetForecaster (pipeline with target transformations)

### Temporal CV
- [x] RollingWindowSplit
- [ ] ExpandingWindowSplit
- [ ] ManualWindowSplit

### Tuning
- [ ] ForecastingGridSearchCV

**Comment by big-o** [comment]
Due to the ThetaForecaster's dependency on SES, #198 also (re-)implements the ExpSmoothingForecaster in the new framework.

**Comment by hweiTPR** [comment]
Is there a way to tune both sklearn & sktime hyperparameters at the same time, when using regressors from sklearn? If not. It could really useful to have it.

**Comment by mloning** [comment]
Hi @hweiTPR, yes you can do something like this: 
```python

regressor_param_grid = {"n_estimators": [100, 200, 300]}
forecaster_param_grid = {"window_length": [10, 12, 14]}
regressor = GridSearchCV(RandomForestRegressor(), param_grid=regressor_param_grid)
forecaster = RecursiveRegressionForecaster(regressor)
cv = SlidingWindowSplitter(initial_window=100)
gscv = ForecastingGridSearchCV(forecaster, cv=cv, param_grid=forecaster_param_grid)
gscv.fit(y_train)

print(gscv.best_params_, gscv.best_forecaster_.regressor_.best_params_)
>>> {'window_length': 12} {'n_estimators': 100}
```

**Comment by hweiTPR** [comment]
That looks great! Would be helpful to have this in the documentation!

**Comment by mloning** [comment]
@hweiTPR I mention it already in the forecasting tutorial below cell 27: https://github.com/alan-turing-institute/sktime/blob/master/examples/01_forecasting.ipynb but without showing the code! Would appreciate a PR if you'd like to add it! 🙂

**Comment by huayicodes** [comment]
@mloning (switching to my personal account) I totally missed that sentence. Happy to contribute. I have another question: what can I put for "score" when using ForecastingGridSearchCV? I tried smape_loss, but it threw me an error.

**Comment by mloning** [comment]
This should work: 
```python
from sktime.forecasting.model_selection import ForecastingGridSearchCV
from sklearn.ensemble import RandomForestRegressor
from sktime.forecasting.compose import RecursiveRegressionForecaster
from sktime.forecasting.model_selection import SlidingWindowSplitter
from sktime.performance_metrics.forecasting import sMAPE
from sklearn.model_selection import GridSearchCV
from sktime.datasets import load_airline

y_train = load_airline()

regressor_param_grid = {"n_estimators": [100, 200, 300]}
forecaster_param_grid = {"window_length": [10, 12, 14]}
regressor = GridSearchCV(RandomForestRegressor(), param_grid=regressor_param_grid)
forecaster = RecursiveRegressionForecaster(regressor)
cv = SlidingWindowSplitter(initial_window=100)
gscv = ForecastingGridSearchCV(forecaster, cv=cv, param_grid=forecaster_param_grid, scoring=sMAPE())
gscv.fit(y_train)
print(gscv.cv_results_)
```
This is a deviation from scikit-learn, as we encapsulated loss functions as classes rather than functions, that makes it easier to attach additional information that's needed when running GSCV. This functionality also still needs some improvements to handle multi-step forecasting horizons.

**Comment by NikeNano** [comment]
Is any one working on an interface to fb prophet? Would be happy to look in to it otherwise!

**Comment by mloning** [comment]
Hi @NikeNano that would be great! It's probably best to create a separate issue for it and link it here!

**Comment by Lovkush-A** [comment]
Been trying to get my head around .reduce_py, and here are some things that I think could be improved.

* The symbol X refers to two different things: 'exogenous inputs' and 'features in the reduced regression problem'. I think we should use explicit variable names, e.g. could have variable 'regression_features'
* The symbol y refers to two different things: input time series and 'target in the reduced regression problem'. I think an alternative could be 'regression_target'
* References to cross-validation are confusing for me. Cross validation ought to be totally separate to learning. I realise that the transformations involved in reduction are closely related to the transformations in cross validation, but it is confusing. Furthermore, the variable 'cv' is not informative and better variable name could be chosen. Something like 'window_indices_creator'.
* My current instinct is we ought to refactor the classes we have into a single class with if loops in them. To add new a new option would then involve just adding another case to three or four functions. Right now, for me to create the multioutput reduction, I need to create a _MultioutputReducer class, and then MultioutputRegression and MultioutputTimeSeriesRegression classes.  However, (if I understood things correctly), multiout should be very similar to direct, and we just need to change one or two lines in the fit and predict stages.

I am new to working on a project of this size, so I do not know the consequences or the history of the decisions. E.g. I do not know how hyper-parameter tuning is affected by all these decisions.

**Comment by mloning** [comment]
I agree with your points: the class structure and variable naming is confusing and hard to understand for new contributors. I'm a bit hesitant about if-statements, mostly because the choice of reduction strategy affects both fit and predict, but I suggest we try it out and see what it looks like.

**Comment by Lovkush-A** [comment]
Copying this comment about possible refactoring of reduction forecaster, by @mloning from issue about exogenous variables.

> Perhaps we should make the reduction compositor more pipeline like, allowing users to pass transformers to specify how to transform the data. In the end, this reduction is a kind of pipeline. This way we could also handle different transformations for different (exogenous) variables.
> 
> Something like this:
> 
> ```python
> forecaster = ReducedRegressionForecaster(regressor=regressor, transformers=[
>     ("t1", SlidingWindowTransformer(window_length, ...), [0]), 
>     ("t2", SlidingWindowTransformer(window_length, ...), [1])
> ], strategy="recursive", ...)
> ```
> 
> with `[0]` referring to the target series and `[1]` to an exogenous variable, together with a new series-to-tabular transformer scitype for the `SlidingWindowTransformer`.

**Comment by yairbeer** [comment]
Can I take gluon-ts interface?

**Comment by fkiraly** [comment]
@yairbeer, yes, that would be nice!

**Comment by aiwalter** [comment]
### Multivariate/vector forecasting
- [x] TransformedTargetForecaster
- [x] ForecastingPipeline
- [x] ForecastingGridSearchCV
- [x] RandomizedGridSearchCV
- [x] MultiplexForecaster
- [ ] make_reduction
- [ ] EnsembleForecaster
- [ ] AutoEnsembleForecaster
- [ ] StackingForecaster

**Comment by TNTran92** [comment]
@yairbeer are you still on gluon-ts?

**Comment by yairbeer** [comment]
@TNTran92 unfortunately I no longer work in TomeSeries and don't have time to contribute at the moment.
Sorry

**Comment by fkiraly** [comment]
that's a shame - could be all yours if you want, @TNTran92!
(if you pick it up, kindly open an issue since it does not seem like there is one)

**Comment by lbventura** [comment]
Checked the forecasters for multivariate time-series and the dynamic factor model is not implemented. Would it be of interest to add it to sktime using the [statsmodels](https://www.statsmodels.org/dev/generated/statsmodels.tsa.statespace.dynamic_factor.DynamicFactor.html) implementation, similar to what was done with the [UnobservedComponents](https://www.sktime.org/en/stable/api_reference/auto_generated/sktime.forecasting.structural.UnobservedComponents.html)?

**Comment by fkiraly** [comment]
@lbventura, I think @Ris-Bali attempted to implement the dynamic factor model but abandoned it here:
https://github.com/alan-turing-institute/sktime/pull/2349

Given that the last activity is one month old, I don´t think he´d mind if you branch off if you try to complete it?

**Comment by Ris-Bali** [comment]
Sorry for leaving it when it was almost done ; I had my exams ; they end this week but I wouldn't mind if @lbventura continues and completed the model.

**Comment by lbventura** [comment]
Great @fkiraly and @Ris-Bali. I'll be tackling it this week and the next, so we can work on it together @Ris-Bali if you want to.

**Comment by Ris-Bali** [comment]
> Great @fkiraly and @Ris-Bali. I'll be tackling it this week and the next, so we can work on it together @Ris-Bali if you want to.

Sure let's work together then .
